### PR TITLE
Add Tapa solver

### PR DESCRIPTION
**Pull request by rhendric**
There are likely cleverer ways to do this but I nerd-sniped myself into implementing the boring approach. Replace with smarter implementations if they're faster.

### CODE REVIEW

File: tapa.py [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+            # Filled cells must form one contiguous region (part 1)
+            s.add(Or(cell.var <= 0, *(
+                And(n.var >= 0, cell.var > n.var) for n in cell.neighbors()
+            )))
+
+    # Filled cells must form one contiguous region (part 2)
+    s.add(Distinct([c.var for c in g.cells]))
+
+    # No 2×2 regions
+    for y in range(g.height - 1):
```
Comment (Zahariel) [inline tapa.py:128]: This is probably easier or at least clearer what you're trying to do with 
```suggestion
    for p in g.points:
        if p.is_outside: continue
        s.add(Not(And([c.var >= 0 for e in [p.edge_left, p.edge_right] for c in e.cells()]))
```
and would be even easier if you added a `p.cells()`. (I just added `is_outside` for one of the ones I did last week.)

File: tapa.py [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+                    ]) for bits in iterate_bitmasks_for_clues(clues, num_bits)
+                ]))
+        else:
+            # Filled cells must form one contiguous region (part 1)
+            s.add(Or(cell.var <= 0, *(
+                And(n.var >= 0, cell.var > n.var) for n in cell.neighbors()
+            )))
+
+    # Filled cells must form one contiguous region (part 2)
+    s.add(Distinct([c.var for c in g.cells]))
```
Comment (Zahariel) [inline tapa.py:125]: This is fine and it works, but I'm curious if it's more or less efficient to relax this constraint to be "at most one 0", and in return tighten the constraint at L120 to be "positive cells must have a non-negative neighbor exactly 1 lower", which is roughly how I did it in Galaxies. Actually I guess you wouldn't even need to tighten up L120; "at most one 0" is actually all you need out of this `Distinct()`.

Comment (Zahariel) [reply tapa.py:125]: But I think tightening L120 will help Z3 by giving it fewer options, probably. Who knows. Z3 performance is completely mysterious.

File: tapa.py (hunk 1 above)
Comment (rhendric) [reply tapa.py:128]: Ehhh, is that clearer? I don't really think of the ‘no 2×2 regions’ restriction in terms of clusters of cells around interior points of the grid. I think my way would generalize more easily to regions of other dimensions and possibly shapes, if that's a variation that ever happens.

File: tapa.py (hunk 2 above)
Comment (rhendric) [reply tapa.py:125]: I just tested and couldn't measure any performance difference doing it one way or the other, but I do know that the `AtLeast`/`AtMost` operators cause Z3 to use a more complex solver, while `Distinct` constraints can be solved in the core SAT solver, so I tend to reach for `Distinct` first. That's pretty much all I know on the matter.

Comment (Zahariel) [reply tapa.py:125]: Huh. I didn't know that about a different solver, although it makes sense. Good to know. I assume `PbEq` and friends similarly require the advanced solver? (I think `AtMost` is secretly implemented in terms of `PbLt` anyway.)

Comment (rhendric) [reply tapa.py:125]: Yeah, I don't know about one being implemented in terms of the other, but my understanding is that they all require the same pseudo-Boolean machinery.

Although that machinery might just be the right thing to use. Last night I only compared the existing code to using `AtMost` and tightening L120, but I just now tried the third possibility of using `AtMost` but leaving L120 alone, and *that* did shave off about 25% of the in-Z3 time. (Which is, admittedly, a fraction of a second—but it's the principle of the thing I guess? It's a good vindication of the `AtMost` approach at least.)

File: tapa.py (hunk 1 above)
Comment (Zahariel) [reply tapa.py:128]: It's funny that we have such totally opposite intuition. It's your code anyway; do it however you like!

**Pull request merged**